### PR TITLE
feat: Identify the connector for a ConnectionConfig by it's unique configuration

### PR DESCRIPTION
**Pull request by hessjcg**
Part of #1226

### CODE REVIEW

File: core/src/test/java/com/google/cloud/sql/core/ConnectorKeyTest.java [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+import static com.google.common.truth.Truth.assertThat;
+
+import com.google.cloud.sql.AuthType;
+import com.google.cloud.sql.ConnectionConfig;
+import java.util.Arrays;
+import org.junit.Test;
+
+public class ConnectorKeyTest {
+
```
Comment (ttosta-google) [inline core/src/test/java/com/google/cloud/sql/core/ConnectorKeyTest.java:27]: Please add tests for the other ConnectionConfig attributes (e.g. ipType)

File: core/src/test/java/com/google/cloud/sql/core/ConnectorKeyTest.java [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    ConnectorKey k2 =
+        new ConnectorKey(
+            new ConnectionConfig.Builder().withCloudSqlInstance("proj:name:inst2").build());
+
+    assertThat(k1).isEqualTo(k2);
+    assertThat(k1.hashCode()).isEqualTo(k2.hashCode());
+  }
+
+  @Test
+  public void adminUrlNotEqual() {
```
Comment (ttosta-google) [inline core/src/test/java/com/google/cloud/sql/core/ConnectorKeyTest.java:55]: Rename to `testKeysNotEqual_withAdminUrlNotEqual` or something similar. Same for the other tests.

File: core/src/main/java/com/google/cloud/sql/core/InternalConnectorRegistry.java [hunk 3]
```diff
@@ -241,7 +241,7 @@ public static void setApplicationName(String applicationName) {
   }
 
   private Connector getConnector(ConnectionConfig config) {
-    return connectors.computeIfAbsent(config.getCloudSqlInstance(), k -> createConnector(config));
+    return connectors.computeIfAbsent(new ConnectorKey(config), k -> createConnector(config));
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/InternalConnectorRegistry.java:244]: In AlloyDB, we used a string to distinguish. Do we prefer one approach over another here?

https://github.com/GoogleCloudPlatform/alloydb-java-connector/blob/7fc16e7e10df096fc0dd226144e348ae4b5c3866/alloydb-jdbc-connector/src/main/java/com/google/cloud/alloydb/ConnectionConfig.java#L78-L91

cc @ttosta-google

File: core/src/test/java/com/google/cloud/sql/core/ConnectorKeyTest.java (hunk 1 above)
Comment (hessjcg) [reply core/src/test/java/com/google/cloud/sql/core/ConnectorKeyTest.java:27]: Done. Also, with better test case names.

File: core/src/test/java/com/google/cloud/sql/core/ConnectorKeyTest.java (hunk 2 above)
Comment (hessjcg) [reply core/src/test/java/com/google/cloud/sql/core/ConnectorKeyTest.java]: Fixed.

File: core/src/main/java/com/google/cloud/sql/core/InternalConnectorRegistry.java (hunk 3 above)
Comment (hessjcg) [reply core/src/main/java/com/google/cloud/sql/core/InternalConnectorRegistry.java:244]: I prefer this approach. Now that we have introduced connection configuration parameters set per-connection configuration like 'targetPrincipal', the InternalConnectorRegistry will need to create a connector for each unique configuration of connector parameters. In this implementation, the InternalConnectorRegistry will automatically create exactly as many connectors as the application needs, usually just 1 connector.

Comment (ttosta-google) [reply core/src/main/java/com/google/cloud/sql/core/InternalConnectorRegistry.java:244]: I like this approach as well. It is cleaner and easy to add new parameters if needed.

**Review by ttosta-google — APPROVED** [review]

File: core/src/main/java/com/google/cloud/sql/core/InternalConnectorRegistry.java (hunk 3 above)
Comment (enocom) [reply core/src/main/java/com/google/cloud/sql/core/InternalConnectorRegistry.java:244]: @ttosta-google let's add this to AlloyDB in that case. Would you file an issue?

Comment (ttosta-google) [reply core/src/main/java/com/google/cloud/sql/core/InternalConnectorRegistry.java:244]: Should we do it as part of https://github.com/GoogleCloudPlatform/alloydb-java-connector/issues/273?

**Pull request merged**